Feature/first cycle isolations

bgmoshe opened the pull request.

bgmoshe committed:
Added new isolation types and init for them.
For now we just put everyone in HOME isolation

bgmoshe committed:
Made agents to be numpy array instead of list.
Made change_agent_relations_by_factor to be able to get dict as input

bgmoshe committed:
Added isolation of weekly and daily first cycle

bgmoshe committed:
Merge branch 'develop' of https://github.com/CoronaHakab/CoronaHakabModel into feature/getting_out_of_isolation

 Conflicts:
	src/corona_hakab_model/manager.py

bgmoshe committed:
Added type hint

bgmoshe committed:
Added type hint

bgmoshe committed:
Fixed time consuming context managing

bgmoshe committed:
Faster way to save calculate the number of isolated of each time

bgmoshe committed:
Symptomatic agents also isolated

bgmoshe committed:
Fixed PR comments

meirhalachmi commented on src/corona_hakab_model/consts.py:
```diff
@@ -219,8 +221,24 @@ class Consts(NamedTuple):
+        True: 1.0,  # 100% sick will obey the isolation.
+        False: .95  # If not sick, 95% to obey isolation
+    }
+    isolation_factor: IsolationFactorsType = {
```
restore the old comment "# reduce agent's relations strength by a factor."

meirhalachmi commented on src/corona_hakab_model/manager.py:
```diff
@@ -158,37 +162,111 @@ def step(self):
+        detected_positive_now = self.tested_positive_vector & \
+                                (self.date_of_last_test == self.current_step)
+        detected_positive_to_isolate = detected_positive_now & (self.agents_in_isolation == IsolationTypes.NONE)
+        detected_positive_indices = np.nonzero(detected_positive_to_isolate)[0]
```
you can use `flatnonzero` instead of `nonzero[0]`

bgmoshe replied:
Done

meirhalachmi commented on src/corona_hakab_model/manager.py:
```diff
@@ -158,37 +162,111 @@ def step(self):
+        detected_positive_indices = np.nonzero(detected_positive_to_isolate)[0]
+        for agent_index in detected_positive_indices:
+            # Get number of days agent is isolated
+            if self.agents_in_isolation[agent_index] != IsolationTypes.NONE:
```
Didn't you generate the detected positive to isolate list using `self.agents_in_isolation == IsolationTypes.NONE`?

bgmoshe replied:
Yeah, I need to remove the previous line. Since you can be in isolation, but it does not mean you wont enter a new kind of isolation

bgmoshe replied:
Is it OK now?

meirhalachmi commented on src/corona_hakab_model/consts.py:
```diff
@@ -219,8 +221,24 @@ class Consts(NamedTuple):
     ]
     should_isolate_positive_detected: bool = False
     isolate_after_num_day: int = 1  # will be in isolation the next day.
```
lousy name.
yes, I know it was there before, but change it :)

bgmoshe replied:
done

bgmoshe replied:
Code I tried, and did not seem to improve:
```
    def progress_isolations(self):
        detected_positive_now = self.tested_positive_vector & \
                                (self.date_of_last_test == self.current_step)
        detected_positive_indices = np.flatnonzero(detected_positive_now)
        days_to_enter_isolation = self.consts.step_to_isolate_dist(size=len(detected_positive_indices))
        list_of_first_circle = [set.union(*[agent_connections.daily_connections for agent_connections in
                                            self.connection_data.connected_ids_by_strength[agent_index].values()])
                                for agent_index in detected_positive_indices]
        agents_in_first_circle = set.union(*list_of_first_circle) if len(list_of_first_circle) > 0 else set()
        daily_steps_to_isolate = self.consts.step_to_isolate_dist(size=len(agents_in_first_circle))
        for i, a in enumerate(agents_in_first_circle):
            if self.step_to_isolate_agent[a] < self.current_step:
                self.step_to_isolate_agent[a] = self.current_step + daily_steps_to_isolate[i]
        for agent_index, steps_to_isolate in \
                zip(detected_positive_indices, days_to_enter_isolation):
            # Get number of days agent is isolated
            if self.agents_in_isolation[agent_index] != IsolationTypes.NONE:
                # TODO: Need to take into account when got out of isolation also
                number_of_days_isolated = max(0,
                                              self.current_step - self.step_to_isolate_agent[agent_index])
            else:
                number_of_days_isolated = 0
            weekly_connection_isolation_ratio = 1 - min(number_of_days_isolated / 7, 1)
            for connected_agents in self.connection_data.connected_ids_by_strength[agent_index].values():
                #daily_steps_to_isolate = self.consts.step_to_isolate_dist(size=len(connected_agents.daily_connections))
                # All of this need to be home isolated
                #for daily_agent_id, daily_days_to_isolate in \
                #        zip(connected_agents.daily_connections, daily_steps_to_isolate):
                    # If were not isolated, and not tested positive now, needs to isolate
                #    if self.agents_in_isolation[daily_agent_id] == IsolationTypes.NONE and \
                #            not self.tested_positive_vector[daily_agent_id]:
                #        if self.step_to_isolate_agent[daily_agent_id] < self.current_step:
                #            self.step_to_isolate_agent[daily_agent_id] = self.current_step + \
                #                                                         daily_days_to_isolate
                agents_to_iterate = list(filter(lambda index: (self.agents_in_isolation[index] == IsolationTypes.NONE)
                                                              and not self.tested_positive_vector[index],
                                                connected_agents.weekly_connections))
                non_sick_agents = self.medical_machine.default_state_upon_infection.agents
                # We want to select which agents to isolate randomly. Starting with sick ones.
                shuffle(agents_to_iterate)
                # We want to first remove agents that are sick
                agents_to_iterate.sort(key=lambda index: self.agents[index] in non_sick_agents, reverse=True)
                how_many_to_isolate = int(round(len(agents_to_iterate) * weekly_connection_isolation_ratio))
                weekly_steps_to_isolate = self.consts.step_to_isolate_dist(size=how_many_to_isolate)
                for i in range(how_many_to_isolate):
                    # If already about to get isolated, do not update it
                    if self.step_to_isolate_agent[agents_to_iterate[i]] > self.current_step:
                        continue
                    self.step_to_isolate_agent[agents_to_iterate[i]] = self.current_step + weekly_steps_to_isolate[i]
            # Isolate the agent
            self.step_to_isolate_agent[agent_index] = self.current_step + steps_to_isolate

        # Isolating symptomatic agents
        for agent in self.new_agents_with_symptoms:
            # If is not getting ready to be isolated, or isolated already, then isolate
            if self.step_to_isolate_agent[agent_index] < self.current_step:
                self.step_to_isolate_agent[agent.index] = self.current_step + self.consts.step_to_isolate_dist
        # TODO: Remove healthy agents from isolation?
        self.isolate_agents()

```

bgmoshe replied:
I cannot remove the loop because I need to go over all the weekly.
Maybe the population was not big enough (20K), maybe on bigger one it can improve, on 20k it did not improve

meirhalachmi replied:
K

meirhalachmi commented on src/corona_hakab_model/manager.py:
```diff
@@ -158,37 +162,111 @@ def step(self):
+                                                              and not self.tested_positive_vector[index],
+                                                connected_agents.weekly_connections))
+                non_sick_agents = self.medical_machine.default_state_upon_infection.agents
+                shuffle(agents_to_iterate)
```
why did you shuffle and then sort?!

bgmoshe replied:
I will add coment, tell me if this is more understandable

meirhalachmi commented on src/corona_hakab_model/manager.py:
```diff
@@ -158,37 +162,111 @@ def step(self):
+                number_of_days_isolated = 0
+            weekly_connection_isolation_ratio = 1 - min(number_of_days_isolated / 7, 1)
+            for connected_agents in self.connection_data.connected_ids_by_strength[agent_index].values():
+                for daily_agent_id in connected_agents.daily_connections:  # All of this need to be home isolated
```
How fast is this entire part? 
I wonder if it won't be any faster to aggregate all the first-cycle agents in a big numpy and then operate on it

bgmoshe replied:
I will check this.
I was not sure myself

meirhalachmi commented on src/corona_hakab_model/manager.py:
```diff
@@ -158,37 +162,111 @@ def step(self):
+                                                connected_agents.weekly_connections))
+                non_sick_agents = self.medical_machine.default_state_upon_infection.agents
+                shuffle(agents_to_iterate)
+                agents_to_iterate.sort(key=lambda index: self.agents[index] in non_sick_agents, reverse=True)
```
why sort?

bgmoshe replied:
This is double comment

meirhalachmi commented on src/corona_hakab_model/manager.py:
```diff
@@ -158,37 +162,111 @@ def step(self):
+        for agent in self.new_agents_with_symptoms:
+            # If is not getting ready to be isolated, or isolated already, then isolate
+            if self.step_to_isolate_agent[agent_index] < self.current_step:
+                self.step_to_isolate_agent[agent.index] = self.current_step + self.consts.isolate_after_num_day
```
I think that symptomatic agents should be isolated immediately, don't you think?

meirhalachmi replied:
They point is that they're isolating themselves cause they feel like crap, not because someone called them and told them that they've been exposed.

bgmoshe replied:
No, I do not agree because of 2 things:
*) Thats the how it was characterized
*) Think about the start of the symptoms, some people want to work, they might say things like "Yeah, I feel a lil bad, but not REALLY bad". Only after number of days of having symptoms they will say: I wanna go home. This might change with isolation policy.

meirhalachmi replied:
Well, it's kinda how it was characterized to me in the phone, but I agree that it's not what they wrote in the paper, so OK

bgmoshe committed:
Added distitbution for days to get into isolation

meirhalachmi commented on src/corona_hakab_model/manager.py:
```diff
@@ -158,37 +162,111 @@ def step(self):
+                agents_to_iterate.sort(key=lambda index: self.agents[index] in non_sick_agents, reverse=True)
+                how_many_to_isolate = round(len(agents_to_iterate) * weekly_connection_isolation_ratio)
+                j = 0
+                for i in range(how_many_to_isolate):
```
How do you enforce that the first "first circle" that will be isolated are the *currently infected* agents?
Not necessarily the symptomatic ones, which should isolate themselves, but anyone that is sick.

bgmoshe replied:
Read the new comment tell me if now it answers

bgmoshe committed:
Merge branch 'develop' of https://github.com/CoronaHakab/CoronaHakabModel into feature/first_cycle_isolations

 Conflicts:
	src/corona_hakab_model/generation/matrix_generator.py

bgmoshe committed:
Answering part of PR1:
*) nonzero(...)[0] -> flatnonzero()
*) detected_positive_to_isolate removed
*) Rename isolate_after_num_day

bgmoshe committed:
Fix comment

bgmoshe committed:
Fix bug where isolated were wrong (did not isolate ones that CURRENTLY git positive test result)

bgmoshe committed:
Remove unused variable

bgmoshe committed:
Merge branch 'develop' of https://github.com/CoronaHakab/CoronaHakabModel into feature/first_cycle_isolations

bgmoshe committed:
Fix some bugs...

bgmoshe committed:
Did a oops

RoyIsland commented on src/corona_hakab_model/detection_model/healthcare.py:
```diff
@@ -3,13 +3,21 @@
     from manager import SimulationManager
 
 
+def _get_current_num_of_tests(current_step, test_location: DetectionSettings):
```
why extract from class?

bgmoshe replied:
Was static
